src/front/glsl/variables.rs

```diff
@@ -0,0 +1,82 @@
+use crate::{
+    Binding, BuiltIn, Expression, GlobalVariable, Handle, ScalarKind, ShaderStage, StorageAccess,
+    StorageClass, Type, TypeInner, VectorSize,
+};
+
+use super::ast::*;
+
+impl Program {
+    pub fn lookup_variable(&mut self, name: &str) -> Option<Handle<Expression>> {
+        let mut expression: Option<Handle<Expression>> = None;
+        match name {
+            "gl_Position" => {
+                if let ShaderStage::Vertex | ShaderStage::Fragment { .. } = self.shader_stage {
+                    let h = self.global_variables.fetch_or_append(GlobalVariable {
+                        name: Some(name.into()),
+                        class: if self.shader_stage == ShaderStage::Vertex {
+                            StorageClass::Output
+                        } else {
+                            StorageClass::Input
+                        },
+                        binding: Some(Binding::BuiltIn(BuiltIn::Position)),
+                        ty: self.types.fetch_or_append(Type {
+                            name: None,
+                            inner: TypeInner::Vector {
+                                size: VectorSize::Quad,
+                                kind: ScalarKind::Float,
+                                width: 4,
+                            },
+                        }),
+                        interpolation: None,
+                        storage_access: StorageAccess::empty(),
+                    });
+                    self.lookup_global_variables.insert(name.into(), h);
+                    let exp = self
+                        .context
+                        .expressions
+                        .append(Expression::GlobalVariable(h));
+                    self.context.lookup_global_var_exps.insert(name.into(), exp);
+
+                    expression = Some(exp);
+                }
+            }
+            "gl_VertexIndex" => {
```

Comment:
These can't be legally used for user variables: `identifiers starting with "gl_" are reserved`
Therefore, instead of filtering the stage out, we should just assume the stage is right (or validate the stages, if you want validation)

Reply:
Ok, I think I'll try to have it error when referencing a `gl_` var not valid for the stage - at least that's what would make most sense to me when thinking as a user passing in a shader :)

Reply:
For ref: what glslang does is error `undeclared identifier`, when trying to reference a `gl_` var in wrong stage. And if trying to declare a user var with `gl_` it errors `identifiers starting with "gl_" are reserved`

Reply:
Looking over it once more, this is how current code behaves (for variable references). Could add the `gl_` prefix check for declarations, but maybe best in separate PR?

Reply:
thinking some more, as a user I'd prefer a better error than just `undeclared identifier`, something like `gl_Position is not available in this stage`. I'll see if I can do something like that, feature flagged with `glsl-validate`